source/tests/system/visa_driver_api_tests.cpp

```diff
@@ -258,6 +270,284 @@ TEST_F(VisaDriverLoopbackTest, WriteWithTermChar_ReadMatches)
+{
+  const int testLength = 5;
+  const int startingOffset = 14;
+  ViUInt16 buffer[testLength] = { 0x0123, 0x0234, 0x00, 0x8675, 0xFEDC };
+  moveOut16(startingOffset, buffer, testLength);
+  for (int i = 0; i < testLength; ++i) {
+    EXPECT_EQ(buffer[i], in16(startingOffset + i * 2));
+  }
+}
+
+TEST_F(VisaRegisterBasedLoopbackTest, DataWrittenWithMoveOut32_In32_ReadsData)
+{
+  const int testLength = 5;
+  const int startingOffset = 16;
+  ViUInt32 buffer[testLength] = { 0x01234567, 0x02340234, 0x00, 0x86754321, 0xC0FFEE };
+  moveOut32(startingOffset, buffer, testLength);
+  for (int i = 0; i < testLength; ++i) {
+    EXPECT_EQ(buffer[i], in32(startingOffset + i * 4));
+  }
+}
+
+TEST_F(VisaRegisterBasedLoopbackTest, DataWrittenWithMoveOut64_In64_ReadsData)
+{
+  const int testLength = 5;
+  const int startingOffset = 16;
+  ViUInt64 buffer[testLength] = { 0x01234567DEADBEEF, 0x02340234, 0x00, 0x867543210ABCDEF, 0xC0FFEE };
+  moveOut64(startingOffset, buffer, testLength);
+  for (int i = 0; i < testLength; ++i) {
+    EXPECT_EQ(buffer[i], in64(startingOffset + i * 8));
+  }
+}
+
+TEST_F(VisaRegisterBasedLoopbackTest, DataWrittenWithOut8_MoveIn8_ReadsData)
+{
+  const int testLength = 5;
+  const int startingOffset = 11;
+  ViUInt8 buffer[testLength] = { 0x01, 0x02, 0x00, 0x03, 0xFE };
+  for (int i = 0; i < testLength; ++i) {
+    out8(startingOffset + i, buffer[i]);
+  }
+  auto readBuffer = moveIn8(startingOffset, testLength);
+  for (int i = 0; i < testLength; ++i) {
+    EXPECT_EQ(readBuffer[i], buffer[i]);
+  }
+}
+
+TEST_F(VisaRegisterBasedLoopbackTest, DataWrittenWithOut16_MoveIn16_ReadsData)
+{
+  const int testLength = 5;
+  const int startingOffset = 14;
+  ViUInt16 buffer[testLength] = { 0x0123, 0x0234, 0x00, 0x8675, 0xFEDC };
+  for (int i = 0; i < testLength; ++i) {
+    out16(startingOffset + i * 2, buffer[i]);
+  }
+  auto readBuffer = moveIn16(startingOffset, testLength);
+  for (int i = 0; i < testLength; ++i) {
+    EXPECT_EQ(readBuffer[i], buffer[i]);
+  }
+}
+
+TEST_F(VisaRegisterBasedLoopbackTest, DataWrittenWithOut32_MoveIn32_ReadsData)
+{
+  const int testLength = 5;
+  const int startingOffset = 16;
+  ViUInt32 buffer[testLength] = { 0x01234567, 0x02340234, 0x00, 0x86754321, 0xC0FFEE };
+  for (int i = 0; i < testLength; ++i) {
+    out32(startingOffset + i * 4, buffer[i]);
+  }
+  auto readBuffer = moveIn32(startingOffset, testLength);
+  for (int i = 0; i < testLength; ++i) {
+    EXPECT_EQ(readBuffer[i], buffer[i]);
+  }
+}
+
+TEST_F(VisaRegisterBasedLoopbackTest, DataWrittenWithOut64_MoveIn64_ReadsData)
+{
+  const int testLength = 5;
+  const int startingOffset = 16;
+  ViUInt64 buffer[testLength] = { 0x01234567DEADBEEF, 0x02340234, 0x00, 0x867543210ABCDEF, 0xC0FFEE };
+  for (int i = 0; i < testLength; ++i) {
+    out64(startingOffset + i * 8, buffer[i]);
+  }
+  auto readBuffer = moveIn64(startingOffset, testLength);
+  for (int i = 0; i < testLength; ++i) {
+    EXPECT_EQ(readBuffer[i], buffer[i]);
+  }
+}
+
```

Comment:
Looks good!
Some ideas for test: Maybe add some tests to do Out64 to write a 64bit data while using In8/In16/In32 to read part of it to check the bit order is aligned with the methods.
Does it allow to read memory whose address is not aligned with byte size? It could be an error case test to read a 64bit data from address 0x1.

Reply:
A few thoughts on why I don't plan to add those here:
- Checking alignment of reads and writes of different sizes is dependent on architecture-specific behavior, specifically big-endian vs. little-endian. Yes we know what architecture our tests run on, but it's not a good idea to depend on that.
- Querying on a misaligned address for a given size will return an error. That's why you see my 8-bit operations start at an odd address but not the others.
- I don't see the value of error cases for the register based operations at all, really. That's for testing the VISA driver. These tests are ensuring that we correctly pass the input parameters from the request packet to the VISA driver and copy the output parameters to the response packet. The error case testing is important in scenarios where our code has to do something special, like handle partial buffer content in the message based operations.

Reply:
`These tests are ensuring that we correctly pass the input parameters from the request packet to the VISA driver and copy the output parameters to the response packet.` 
Good point!